blockchain/message_pool/src/msgpool/selection.rs

```diff
@@ -0,0 +1,592 @@
+        }
+        // 1. Create a list of dependent message chains with maximal gas reward per limit consumed
+        let mut chains = Vec::new();
+        for (actor, mset) in pending.into_iter() {
+            chains.append(
+                &mut create_message_chains(&self.api, &actor, &mset, &base_fee, &ts).await?,
+            );
+        }
+
+        // 2. Sort the chains
+        chains.sort_by(|a, b| b.compare(&a));
+        if !allow_negative_chains(cur_ts.epoch())
+            && !chains.is_empty()
+            && chains[0].curr().gas_perf < 0.0
+        {
+            return Ok(result);
+        }
+        // 3. Merge the head chains to produce the list of messages selected for inclusion, subject to
+        //    the block gas limit.
+        let mut last = chains.len();
+        for i in 0..chains.len() {
+            if !allow_negative_chains(cur_ts.epoch()) && chains[i].curr().gas_perf < 0.0 {
+                break;
+            }
+
+            if chains[i].curr().gas_limit <= gas_limit {
+                gas_limit -= chains[i].curr().gas_limit;
+                result.append(&mut chains[i].curr().msgs.clone());
+                continue;
+            }
+            last = i;
+            break;
+        }
+        // 4. We have reached the edge of what we can fit wholesale; if we still have available gasLimit
+        // to pack some more chains, then trim the last chain and push it down.
+        // Trimming invalidates subsequent dependent chains so that they can't be selected as their
+        // dependency cannot be (fully) included.
+        // We do this in a loop because the blocker might have been inordinately large and we might
+        // have to do it multiple times to satisfy tail packing.
+        'tail_loop: while gas_limit >= min_gas && last < chains.len() {
+            chains[last].trim(gas_limit, &base_fee, allow_negative_chains(cur_ts.epoch()));
+            if chains[last].curr().valid {
+                for i in last..(chains.len() - 1) {
+                    if chains[i].compare(&chains[i + 1]) == Ordering::Greater {
+                        break;
+                    }
+                    chains.swap(i, i + 1);
+                }
+            }
+
+            // select the next (valid and fitting) chain for inclusion
+            for i in last..chains.len() {
+                if !chains[i].curr().valid {
+                    continue;
+                }
+
+                // if gas_perf < 0 then we have no more profitable chains
+                if !allow_negative_chains(cur_ts.epoch()) && chains[i].curr().gas_perf < 0.0 {
+                    break 'tail_loop;
+                }
+
+                // does it fit in the block?
+                if chains[i].curr().gas_limit <= gas_limit {
+                    gas_limit -= chains[i].curr().gas_limit;
+                    result.append(&mut chains[i].curr_mut().msgs);
+                    continue;
+                }
+                last += i;
```

Comment:
Should this be `= i`?

Reply:
No I dont think so.. why do you say that?

Reply:
It seems that way because `i` is an index into the `chains` vector, why is it added to `last`? If the `.enumerate()` call came after the `.skip(last)` call then `+= i` would make sense to me

Reply:
ohhhhhhhh wow thats a really really good catch dude... enumerate should come after skip